File: engine/src/flutter/lib/gpu/surface.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/lib/gpu/surface.h"

#include <cstdint>

#include "flutter/lib/gpu/formats.h"
#include "flutter/lib/ui/painting/image.h"
#include "fml/logging.h"
#include "impeller/core/allocator.h"
#include "tonic/converter/dart_converter.h"

#if IMPELLER_SUPPORTS_RENDERING
#include "impeller/display_list/dl_image_impeller.h"  // nogncheck
#endif

namespace flutter {
namespace gpu {

IMPLEMENT_WRAPPERTYPEINFO(flutter_gpu, Surface);

namespace {

constexpr int64_t kMaxInternalTextureReferences = 3;

}  // namespace

Surface::TextureRecord::TextureRecord(
    std::shared_ptr<impeller::Texture> texture,
    sk_sp<DlImage> image,
    impeller::ISize size,
    impeller::PixelFormat format)
    : texture(std::move(texture)),
      image(std::move(image)),
      size(size),
      format(format) {}

Surface::Surface(std::shared_ptr<impeller::Context> context,
                 impeller::ISize size,
                 impeller::PixelFormat format)
    : context_(std::move(context)), size_(size), format_(format) {}

Surface::~Surface() = default;

std::shared_ptr<Surface::TextureRecord> Surface::CreateTextureRecord() const {
#if !IMPELLER_SUPPORTS_RENDERING
  FML_LOG(ERROR) << "Flutter GPU surfaces require Impeller rendering support.";
  return nullptr;
#else
  impeller::TextureDescriptor desc;
  desc.storage_mode = impeller::StorageMode::kDevicePrivate;
  desc.size = size_;
  desc.format = format_;
  desc.sample_count = impeller::SampleCount::kCount1;
  desc.type = impeller::TextureType::kTexture2D;
  desc.mip_count = 1;
  desc.usage = {};
  desc.usage |= impeller::TextureUsage::kRenderTarget;
  desc.usage |= impeller::TextureUsage::kShaderRead;

  auto texture = context_->GetResourceAllocator()->CreateTexture(desc, true);
  if (!texture) {
    FML_LOG(ERROR) << "Failed to create Flutter GPU surface texture.";
    return nullptr;
  }
  texture->SetCoordinateSystem(
      impeller::TextureCoordinateSystem::kRenderToTexture);

  auto image = impeller::DlImageImpeller::Make(texture);
  if (!image) {
    FML_LOG(ERROR) << "Failed to create Flutter GPU surface image.";
    return nullptr;
  }

  return std::make_shared<TextureRecord>(std::move(texture), std::move(image),
                                         size_, format_);
#endif
}

bool Surface::IsReusable(const std::shared_ptr<TextureRecord>& record,
                         size_t index) const {
  if (current_index_.has_value() && current_index_.value() == index) {
    return false;
  }
  return record && record->size.width == size_.width &&
         record->size.height == size_.height && record->format == format_ &&
         !record->acquired && !record->producer_pending.load() &&
         record->image && record->image->unique() &&
         static_cast<int64_t>(record->texture.use_count()) <=
             kMaxInternalTextureReferences;
}

void Surface::PruneTextureRecords() {
  for (size_t i = 0; i < records_.size(); i++) {
    auto& record = records_[i];
    if (!record) {
      continue;
    }

    const bool is_current =
        current_index_.has_value() && current_index_.value() == i;
    const bool matches_surface = record->size.width == size_.width &&
                                 record->size.height == size_.height &&
                                 record->format == format_;
    // The surface record, its image, and the Dart texture wrapper handed out
    // for the completed frame may all still reference an otherwise
    // unreferenced texture.
    const bool has_external_references =
        !record->image || !record->image->unique() ||
        static_cast<int64_t>(record->texture.use_count()) >
            kMaxInternalTextureReferences;

    if (!is_current && !matches_surface && !record->acquired &&
        !record->producer_pending.load() && !has_external_references) {
      record.reset();
    }
  }

  while (!records_.empty() && !records_.back()) {
    records_.pop_back();
  }
}

int Surface::AcquireNextFrame(Dart_Handle texture_wrapper) {
  std::optional<size_t> available_index;
  size_t index = 0;
  for (const auto& record : records_) {
    if (!record && !available_index.has_value()) {
      available_index = index;
    }
    if (IsReusable(record, index)) {
      record->acquired = true;
      auto texture = fml::MakeRefCounted<Texture>(record->texture);
      texture->AssociateWithDartWrapper(texture_wrapper);
      return static_cast<int>(index);
    }
    index++;
  }

  auto record = CreateTextureRecord();
  if (!record) {
    return -1;
  }
  record->acquired = true;

  if (available_index.has_value()) {
    records_[available_index.value()] = record;
    index = available_index.value();
  } else {
    records_.push_back(record);
    index = records_.size() - 1u;
  }

  auto texture = fml::MakeRefCounted<Texture>(record->texture);
  texture->AssociateWithDartWrapper(texture_wrapper);
  return static_cast<int>(index);
}

Dart_Handle Surface::CreateImage(const sk_sp<DlImage>& image) const {
  if (!image) {
    return Dart_Null();
  }
  auto canvas_image = CanvasImage::Create();
  canvas_image->set_image(image);
  return canvas_image->CreateOuterWrapping();
}

Dart_Handle Surface::PresentFrame(size_t texture_index,
                                  CommandBuffer& command_buffer) {
  if (texture_index >= records_.size()) {
    return tonic::ToDart("SurfaceFrame does not belong to this GpuSurface.");
  }

  auto record = records_[texture_index];
  if (!record || !record->acquired) {
    return tonic::ToDart(
        "SurfaceFrame has already been presented or discarded.");
  }

  record->producer_pending.store(true);
  if (!command_buffer.AddCompletionCallback(
          [record](impeller::CommandBuffer::Status status) {
            (void)status;
            record->producer_pending.store(false);
          })) {
    record->producer_pending.store(false);
    return tonic::ToDart(
        "SurfaceFrame.present must be called before submitting the command "
        "buffer.");
  }

  record->acquired = false;
  current_index_ = texture_index;
  PruneTextureRecords();
  return CreateImage(record->image);
}

void Surface::DiscardFrame(size_t texture_index) {
  if (texture_index >= records_.size()) {
    return;
  }
  auto record = records_[texture_index];
  if (record) {
    record->acquired = false;
  }
}

Dart_Handle Surface::GetCurrentImage() const {
  if (!current_index_.has_value() ||
      current_index_.value() >= records_.size()) {
    return Dart_Null();
  }
  auto record = records_[current_index_.value()];
  if (!record) {
    return Dart_Null();
  }
  return CreateImage(record->image);
}

std::optional<std::string> Surface::Resize(impeller::ISize size) {
  for (const auto& record : records_) {
    if (record && record->acquired) {
      return "GpuSurface.resize cannot be called while a SurfaceFrame is "
             "acquired.";
    }
  }
  size_ = size;
  PruneTextureRecords();
  return std::nullopt;
}

size_t Surface::GetBackingTextureCount() const {
  size_t count = 0;
  for (const auto& record : records_) {
    if (record) {
      count++;
    }
  }
  return count;
}

}  // namespace gpu
}  // namespace flutter

//----------------------------------------------------------------------------
/// Exports
///

Dart_Handle InternalFlutterGpu_Surface_Initialize(
    Dart_Handle wrapper,
    flutter::gpu::Context* gpu_context,
    int width,
    int height,
    int format) {
  if (width <= 0 || height <= 0) {
    return tonic::ToDart("GpuSurface dimensions must be greater than zero.");
  }

  auto pixel_format = flutter::gpu::ToImpellerPixelFormat(format);
  if (pixel_format == impeller::PixelFormat::kUnknown) {
    return tonic::ToDart("Unsupported GpuSurface pixel format.");
  }

  auto res = fml::MakeRefCounted<flutter::gpu::Surface>(
      gpu_context->GetContextShared(), impeller::ISize{width, height},
      pixel_format);
  res->AssociateWithDartWrapper(wrapper);
  return Dart_Null();
}

int InternalFlutterGpu_Surface_AcquireNextFrame(flutter::gpu::Surface* wrapper,
                                                Dart_Handle texture_wrapper) {
  return wrapper->AcquireNextFrame(texture_wrapper);
}

Dart_Handle InternalFlutterGpu_Surface_PresentFrame(
    flutter::gpu::Surface* wrapper,
    int texture_index,
    flutter::gpu::CommandBuffer* command_buffer) {
  if (texture_index < 0) {
    return tonic::ToDart("Invalid SurfaceFrame texture index.");
  }
  return wrapper->PresentFrame(static_cast<size_t>(texture_index),
                               *command_buffer);
}

void InternalFlutterGpu_Surface_DiscardFrame(flutter::gpu::Surface* wrapper,
                                             int texture_index) {
  if (texture_index < 0) {
    return;
  }
  wrapper->DiscardFrame(static_cast<size_t>(texture_index));
}

Dart_Handle InternalFlutterGpu_Surface_GetCurrentImage(
    flutter::gpu::Surface* wrapper) {
  return wrapper->GetCurrentImage();
}

Dart_Handle InternalFlutterGpu_Surface_Resize(flutter::gpu::Surface* wrapper,
                                              int width,
                                              int height) {
  if (width <= 0 || height <= 0) {
    return tonic::ToDart("GpuSurface dimensions must be greater than zero.");
  }
  auto error = wrapper->Resize(impeller::ISize{width, height});
  if (error.has_value()) {
    return tonic::ToDart(error.value());
  }
  return Dart_Null();
}

int InternalFlutterGpu_Surface_GetBackingTextureCount(
    flutter::gpu::Surface* wrapper) {
  return static_cast<int>(wrapper->GetBackingTextureCount());
}

File: engine/src/flutter/lib/gpu/lib/src/surface.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

part of flutter_gpu;

/// The result of presenting a [GpuSurfaceFrame].
///
/// Image surfaces always report [success]. The other values exist for
/// destinations whose presentation can fail or require reconfiguration, so that
/// one calling pattern is correct for every surface type.
enum GpuPresentStatus {
  /// The frame was presented and no reconfiguration is needed.
  success,

  /// The frame was presented, but the surface should be reconfigured before the
  /// next frame.
  suboptimal,

  /// The surface is out of date and must be reconfigured before presenting
  /// again.
  outOfDate,
}

/// A presentable Flutter GPU render target.
///
/// A surface vends [GpuSurfaceFrame]s. Each frame exposes a color texture to
/// render into and is published with [GpuSurfaceFrame.present]. The destination
/// of a surface is fixed when the surface is created.
///
/// Use [GpuContext.createImageSurface] to create an image surface that Flutter
/// draws as a [ui.Image].
sealed class GpuSurface {
  /// Acquires the next frame to render into, or null if no frame is available.
  ///
  /// Image surfaces always return a frame. The nullable result exists for
  /// destinations that may skip a frame, such as a throttled window swapchain.
  GpuSurfaceFrame? acquireNextFrame();
}

/// A writable frame acquired from a [GpuSurface].
///
/// A frame is a temporary write lease for [colorTexture]. Record the final
/// color writes into that texture, then call [present]. If the frame will not
/// be presented, call [discard] so the surface can reuse its backing texture.
///
/// Call exactly one of [present] or [discard] for each acquired frame. Calling
/// [discard] after [present] is a no-op, and calling [present] after [discard]
/// throws.
sealed class GpuSurfaceFrame {
  /// The color texture to use as the final render target for this frame.
  ///
  /// This texture is render-target and shader-read capable. It is writable
  /// until [present] or [discard] is called. After that point, applications
  /// must not render into it again.
  Texture get colorTexture;

  /// Publishes this frame and returns the presentation status.
  ///
  /// The [commandBuffer] must be the command buffer that contains the final
  /// writes to [colorTexture]. Call this method after recording those writes
  /// and before submitting [commandBuffer].
  ///
  /// Calling this method does not submit [commandBuffer]. You must submit it for
  /// the presented frame to be rendered.
  ///
  /// For an image surface, the published image is available from
  /// [GpuImageSurface.currentImage].
  GpuPresentStatus present(CommandBuffer commandBuffer);

  /// Releases this frame without presenting it.
  ///
  /// Use this when rendering is abandoned after acquiring a frame. Discarding a
  /// frame does not change [GpuImageSurface.currentImage].
  void discard();
}

/// A pool of presentable Flutter GPU render targets drawn by Flutter as a
/// [ui.Image].
///
/// An image surface is useful when GPU content is rendered repeatedly and then
/// drawn by Flutter as a [ui.Image]. It owns the final color textures that may
/// be sampled by Flutter after a frame is presented. The renderer is still
/// responsible for ordinary intermediate render targets such as depth, stencil,
/// multisample, shadow, or post-processing textures.
///
/// The surface chooses how many backing textures to retain. Applications
/// should acquire a frame when they are ready to render and let the surface
/// decide whether an existing texture can be reused or a new texture is needed.
///
/// To render a frame, call [acquireNextFrame], render into the returned
/// [GpuSurfaceFrame.colorTexture], call [GpuSurfaceFrame.present] with the
/// command buffer that contains the final writes, submit that command buffer,
/// and then draw [currentImage].
final class GpuImageSurface extends NativeFieldWrapperClass1
    implements GpuSurface {
  GpuImageSurface._(this._gpuContext, int width, int height, PixelFormat format)
    : _width = width,
      _height = height,
      _format = format {
    _checkDimensions(width, height);
    final String? error = _initialize(_gpuContext, width, height, format.index);
    if (error != null) {
      throw Exception(error);
    }
  }

  final GpuContext _gpuContext;

  int _width;
  int _height;
  PixelFormat _format;

  // The command buffer used to present the most recent frame. It is retained
  // only to detect the case where a frame is presented but its command buffer
  // is never submitted, which would leave the presented image blank forever.
  CommandBuffer? _pendingPresentCommandBuffer;

  /// The width, in pixels, of frames acquired from this surface.
  int get width => _width;

  /// The height, in pixels, of frames acquired from this surface.
  int get height => _height;

  /// The color format of frames acquired from this surface.
  PixelFormat get format => _format;

  /// The number of backing textures currently retained by this surface.
  ///
  /// This value is intended for diagnostics and tests. Applications should not
  /// branch on a particular count because the surface may grow the pool when
  /// Flutter or the GPU is still using older presented frames.
  int get debugBackingTextureCount => _getBackingTextureCount();

  /// The most recently presented image, or null if no frame has been presented.
  ///
  /// Each call returns a [ui.Image] handle for the current presented frame. The
  /// image remains valid even if later calls to [acquireNextFrame] return a
  /// different backing texture.
  ui.Image? get currentImage {
    final Object? result = _getCurrentImage();
    if (result == null) {
      return null;
    }
    return result as ui.Image;
  }

  /// Changes the size of future frames acquired from this surface.
  ///
  /// Existing images returned by [currentImage] remain valid. The next
  /// [acquireNextFrame] call returns a texture with the new dimensions. This
  /// method throws if a frame is currently acquired and has not been presented
  /// or discarded.
  void resize(int width, int height) {
    _checkDimensions(width, height);
    final String? error = _resize(width, height);
    if (error != null) {
      throw StateError(error);
    }
    _width = width;
    _height = height;
  }

  /// Acquires a color texture that can be used as the final render target.
  ///
  /// The returned [GpuImageSurfaceFrame] owns the write lease for its
  /// [GpuSurfaceFrame.colorTexture]. Renderers should write their final color
  /// result into that texture and then call [GpuSurfaceFrame.present].
  ///
  /// The surface will not return a backing texture that is still the current
  /// image, is waiting on producer GPU work, or is still referenced by Flutter.
  /// If no retained backing texture is available, the surface allocates another
  /// one.
  @override
  GpuImageSurfaceFrame acquireNextFrame() {
    _checkPreviousPresentSubmitted();
    final Texture texture = Texture._surface(
      _gpuContext,
      _format,
      _width,
      _height,
    );
    final int textureIndex = _acquireNextFrame(texture);
    if (textureIndex < 0) {
      throw Exception('Failed to acquire a Flutter GPU surface frame.');
    }
    texture._valid = true;
    return GpuImageSurfaceFrame._(this, textureIndex, texture);
  }

  void _presentFrame(GpuImageSurfaceFrame frame, CommandBuffer commandBuffer) {
    final Object? result = _presentFrameNative(
      frame._textureIndex,
      commandBuffer,
    );
    if (result is String) {
      throw Exception(result);
    }
    _pendingPresentCommandBuffer = commandBuffer;
  }

  // Verifies that the command buffer used to present the previous frame was
  // submitted. A presented frame is not rendered until its command buffer is
  // submitted, so a forgotten submit would otherwise silently leave the
  // surface's image blank.
  void _checkPreviousPresentSubmitted() {
    final CommandBuffer? commandBuffer = _pendingPresentCommandBuffer;
    if (commandBuffer != null && !commandBuffer.submitted) {
      throw StateError(
        'The command buffer used to present the previous GpuSurfaceFrame was '
        'never submitted. Call commandBuffer.submit() after '
        'GpuSurfaceFrame.present() so the presented frame is rendered.',
      );
    }
    _pendingPresentCommandBuffer = null;
  }

  void _discardFrame(GpuImageSurfaceFrame frame) {
    _discardFrameNative(frame._textureIndex);
  }

  static void _checkDimensions(int width, int height) {
    if (width <= 0) {
      throw ArgumentError.value(width, 'width', 'Must be greater than zero.');
    }
    if (height <= 0) {
      throw ArgumentError.value(height, 'height', 'Must be greater than zero.');
    }
  }

  @Native<Handle Function(Handle, Pointer<Void>, Int, Int, Int)>(
    symbol: 'InternalFlutterGpu_Surface_Initialize',
  )
  external String? _initialize(
    GpuContext gpuContext,
    int width,
    int height,
    int format,
  );

  @Native<Int Function(Pointer<Void>, Handle)>(
    symbol: 'InternalFlutterGpu_Surface_AcquireNextFrame',
  )
  external int _acquireNextFrame(Texture texture);

  @Native<Handle Function(Pointer<Void>, Int, Pointer<Void>)>(
    symbol: 'InternalFlutterGpu_Surface_PresentFrame',
  )
  external Object? _presentFrameNative(
    int textureIndex,
    CommandBuffer commandBuffer,
  );

  @Native<Void Function(Pointer<Void>, Int)>(
    symbol: 'InternalFlutterGpu_Surface_DiscardFrame',
  )
  external void _discardFrameNative(int textureIndex);

  @Native<Handle Function(Pointer<Void>)>(
    symbol: 'InternalFlutterGpu_Surface_GetCurrentImage',
  )
  external Object? _getCurrentImage();

  @Native<Handle Function(Pointer<Void>, Int, Int)>(
    symbol: 'InternalFlutterGpu_Surface_Resize',
  )
  external String? _resize(int width, int height);

  @Native<Int Function(Pointer<Void>)>(
    symbol: 'InternalFlutterGpu_Surface_GetBackingTextureCount',
  )
  external int _getBackingTextureCount();
}

/// A writable frame acquired from a [GpuImageSurface].
final class GpuImageSurfaceFrame implements GpuSurfaceFrame {
  GpuImageSurfaceFrame._(this._surface, this._textureIndex, this.colorTexture);

  final GpuImageSurface _surface;
  final int _textureIndex;
  bool _isActive = true;

  @override
  final Texture colorTexture;

  /// Publishes this frame as the current image for the surface.
  ///
  /// The [commandBuffer] must be the command buffer that contains the final
  /// writes to [colorTexture]. Call this method after recording those writes
  /// and before submitting [commandBuffer].
  ///
  /// Calling this method does not submit [commandBuffer]. It registers the
  /// frame with [commandBuffer] so the surface can use command buffer
  /// completion to decide when this texture can be considered for reuse.
  ///
  /// You must submit [commandBuffer] for the presented frame to be rendered.
  /// Until it is submitted, [GpuImageSurface.currentImage] will not contain
  /// this frame's contents, and the backing texture remains unavailable for
  /// future frames. If a frame is presented but its command buffer is never
  /// submitted, the next [GpuImageSurface.acquireNextFrame] throws a
  /// [StateError].
  ///
  /// The published image is available from [GpuImageSurface.currentImage].
  @override
  GpuPresentStatus present(CommandBuffer commandBuffer) {
    if (!_isActive) {
      throw StateError(
        'GpuSurfaceFrame has already been presented or discarded.',
      );
    }
    _surface._presentFrame(this, commandBuffer);
    _isActive = false;
    colorTexture._valid = false;
    return GpuPresentStatus.success;
  }

  /// Releases this frame without presenting it.
  ///
  /// Use this when rendering is abandoned after acquiring a frame. Discarding a
  /// frame does not change [GpuImageSurface.currentImage].
  @override
  void discard() {
    if (!_isActive) {
      return;
    }
    _isActive = false;
    colorTexture._valid = false;
    _surface._discardFrame(this);
  }
}
